pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -27,100 +29,164 @@ import (
 	client client.Client
 	scheme *runtime.Scheme
 }
 
-// Reconcile reads that state of the cluster for a NSTemplateTier object and makes changes based on the state read
-// and what is in the NSTemplateTier.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNSTemplateTier) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+const (
+	// MaxPoolSize the maximum number of TemplateUpdateRequest resources that can exist at the same time
+	MaxPoolSize = 5
+)
+
+// Reconcile takes care of:
+// - inserting a new entry in the `status.updates` (and cleaning the 'failedAccounts` in the previous one)
+// - creating and delete the TemplateUpdateRequest to update the MasterUserRecord associated with this tier
+// - updating the `Failed` counter in the `status.updates` when a MasterUserRecord failed to update
+// - setting the `completionTime` when all MasterUserRecord have been processed
+func (r *NSTemplateTierReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
-	logger.Info("Reconciling NSTemplateTier")
 
-	// Fetch the NSTemplateTier tier
+	// fetch the NSTemplateTier tier
 	tier := &toolchainv1alpha1.NSTemplateTier{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, tier)
-	if err != nil {
+	if err := r.client.Get(context.TODO(), request.NamespacedName, tier); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("NSTemplateTier not found")
+			logger.Info("NSTemplateTier not found")
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		logger.Error(err, "unable to get the current NSTemplateTier tier")
 		return reconcile.Result{}, err
 	}
-	if err := r.ensureTemplateUpdateRequest(logger, tier); err != nil {
+
+	// create a new entry in the `status.history`
+	if added, err := r.ensureStatusUpdateRecord(logger, tier); err != nil {
+		logger.Error(err, "unable to insert a new entry in status.updates after NSTemplateTier changed.")
+		return reconcile.Result{}, err
+	} else if added {
+		logger.Info("Requeing after adding a new entry in tier.status.updates")
+		return reconcile.Result{Requeue: true}, nil
+	}
+	if done, err := r.ensureTemplateUpdateRequest(logger, tier); err != nil {
 		logger.Error(err, "unable to ensure TemplateRequestUpdate resource after NSTemplateTier changed.")
 		return reconcile.Result{}, err
+	} else if done {
+		log.Info("No MasterUserRecord to update. Setting the completion timestamp immediately")
+		if err := r.markUpdateRecordAsCompleted(tier); err != nil {
+			logger.Error(err, "unable to mark latest status.update as complete.")
+			return reconcile.Result{}, err
+		}
+		logger.Info("No MasterUserRecord to update")
```

Comment:
Already logged in line 138

Reply:
ah yes, indeed. Thanks! Removed in 0621d5b